### PR TITLE
port to node addon api

### PR DESCRIPTION
**Pull request by geovie**
This PR ports node-usb to use [node addon api](https://nodejs.github.io/node-addon-examples/about/what/) instead of [nan](https://github.com/nodejs/nan).

The migration was mostly done using the [migration tool](https://nodejs.github.io/node-addon-examples/getting-started/migration/#run-the-conversion-tool) and some manual adjustments.

One upside of this, besides of using a more recent api, is that node addon api allows [prebuilds](https://nodejs.github.io/node-addon-examples/build-tools/prebuild/) to be build once and be used by various node (and electron) versions which support the necessary node addon api (version 3 is used in this PR which is supported by most [node versions](https://nodejs.org/api/n-api.html#n_api_n_api_version_matrix)).

### CODE REVIEW

File: src/device.cc [hunk 1]
```diff
@@ -29,28 +31,25 @@ std::map<libusb_device*, Device*> Device::byPtr;
...
-		Local<FunctionTemplate> constructorHandle = Nan::New<v8::FunctionTemplate>(device_constructor);
-		Local<Value> argv[1] = { EXTERNAL_NEW(new Device(dev)) };
-		Local<Object> obj = Nan::NewInstance(Nan::GetFunction(constructorHandle).ToLocalChecked(), 1, argv).ToLocalChecked();
+		Napi::Object obj = Device::constructor.New({ Napi::External<libusb_device>::New(env, dev) });
 		return obj;
 	}
 }
 
-static NAN_METHOD(deviceConstructor) {
+Napi::Value Device::Constructor(const Napi::CallbackInfo& info) {
```
Comment (geovie) [inline src/device.cc:44]: This could be moved to the C++ constructor but was kept to minimize the diff

File: src/transfer.cc [hunk 2]
```diff
@@ -22,7 +23,8 @@ Transfer::~Transfer(){
 }
 
 // new Transfer(device, endpointAddr, type, timeout)
-NAN_METHOD(Transfer_constructor) {
+Napi::Value Transfer::Constructor(const Napi::CallbackInfo& info) {
```
Comment (geovie) [inline src/transfer.cc:26]: This could be moved to the C++ constructor but was kept to minimize the diff

File: src/node_usb.h [hunk 3]
```diff
@@ -9,71 +9,88 @@
...
 
-	static void Init(Local<Object> exports);
-	static Local<Object> get(libusb_device* handle);
+	static Napi::Object Init(Napi::Env env, Napi::Object exports);
+	static Napi::Object get(napi_env env, libusb_device* handle);
 
-	inline void ref(){Ref();}
-	inline void unref(){Unref();}
+	inline void ref(){ refs_ = Ref();}
+	inline void unref(){ refs_ = Unref();}
```
Comment (geovie) [inline src/node_usb.h:29]: I'm unsure about this... nan exposed the refs_ count directly

**Comment by geovie** [comment]
Pushed another commit which ports the pieces where libuv was used to use napi too, as libuv doesn't provide an [ABI stability guarantee](https://nodejs.org/api/n-api.html#n_api_implications_of_abi_stability) across major node versions.

Since `ThreadSafeFunction` is used the minimum NAPI_VERSION is now 4.

**Comment by kevinmehall** [comment]
Thanks @geovie! This looks great and is much-needed. I've been hoping someone would do this for a while.

What changes are required for CI / release scripts? I assume prebuild can be configured to build only one binary library per platform instead of per version?

Unfortunately, I'm not really using or maintaining this library anymore and don't have time to review / test / release this right now. If you're interested in helping, I'd be happy to give you access. See #401.

**Comment by geovie** [comment]
@kevinmehall For the CI the only thing that would change is that instead of specifying targets you would use the napi runtime `prebuild --runtime napi`.

And some small adjustments to the `package.json` and `bindings.gyp` files as outlined [here](https://nodejs.github.io/node-addon-examples/build-tools/prebuild/)

Happy to create a PR with those changes too, but first this PR should land, so if anyone could review/test this that would be great.

**Comment by thegecko** [comment]
@geovie I've checked out your branch and tried it locally.

All of the device tests pass, but there is an error when closing the device at the end:

```bash
  Module
    ✓ should describe basic constants
    ✓ should handle abuse without crashing
    setDebugLevel
      ✓ should throw when passed invalid args
      ✓ should succeed with good args

  getDeviceList
    ✓ should return at least one device

  findByIds
    ✓ should return an array with length > 0

  Device
    ✓ should have sane properties
    ✓ should have a deviceDescriptor property
    ✓ should have a configDescriptor property
    ✓ should open
    ✓ gets string descriptors
    control transfer
      ✓ should OUT transfer when the IN bit is not set
      ✓ should fail when bmRequestType doesn't match buffer / length
      ✓ should IN transfer when the IN bit is set
      ✓ should signal errors
    Interface
      ✓ should have one interface
      ✓ should be the same as the interfaceNo 0
      IN endpoint
        ✓ should be able to get the endpoint
        ✓ should be able to get the endpoint by address
        ✓ should have the IN direction flag
        ✓ should have a descriptor
        ✓ should fail to write
        ✓ should support read
        ✓ times out
        ✓ polls the device (1685ms)
      OUT endpoint
        ✓ should be able to get the endpoint
        ✓ should be able to get the endpoint by address
        ✓ should have the OUT direction flag
        ✓ should support write
        ✓ times out
      1) "after all" hook for "should be the same as the interfaceNo 0"


  30 passing (2s)
  1 failing

  1) Device
       Interface
         "after all" hook for "should be the same as the interfaceNo 0":
     TypeError: self.device.__releaseInterface is not a function
      at next (usb.js:311:15)
      at Interface.release (usb.js:295:3)
      at Context.<anonymous> (test/usb.coffee:187:10)
      at processImmediate (internal/timers.js:461:21)

```

Any idea what that could be?

> For the CI the only thing that would change is that instead of specifying targets you would use the napi runtime prebuild --runtime napi.

Would be great to see the [prebuild file](https://github.com/tessel/node-usb/blob/master/.github/workflows/prebuild.yml#L35) updated accordingly.

**Comment by geovie** [comment]
@thegecko Should be fixed now. Could you please test it again 🙏 ?     
Should I add the prebuild changes to this PR or create a follow up?

**Comment by thegecko** [comment]
> Should I add the prebuild changes to this PR or create a follow up?

I think it makes sense to include the changes in this PR.

**Comment by thegecko** [comment]
> Could you please test it again

I can confirm all device tests now pass (checked with Node 10 and Node 12) 🎉 

```bash
> usb@1.6.3 full-test /Users/thegecko/Projects/node-usb
> mocha --require coffeescript/register test/*



  Module
    ✓ should describe basic constants
    ✓ should handle abuse without crashing
    setDebugLevel
      ✓ should throw when passed invalid args
      ✓ should succeed with good args

  getDeviceList
    ✓ should return at least one device

  findByIds
    ✓ should return an array with length > 0

  Device
    ✓ should have sane properties
    ✓ should have a deviceDescriptor property
    ✓ should have a configDescriptor property
    ✓ should open
    ✓ gets string descriptors
    control transfer
      ✓ should OUT transfer when the IN bit is not set
      ✓ should fail when bmRequestType doesn't match buffer / length
      ✓ should IN transfer when the IN bit is set
      ✓ should signal errors
    Interface
      ✓ should have one interface
      ✓ should be the same as the interfaceNo 0
      IN endpoint
        ✓ should be able to get the endpoint
        ✓ should be able to get the endpoint by address
        ✓ should have the IN direction flag
        ✓ should have a descriptor
        ✓ should fail to write
        ✓ should support read
        ✓ times out
        ✓ polls the device (1674ms)
      OUT endpoint
        ✓ should be able to get the endpoint
        ✓ should be able to get the endpoint by address
        ✓ should have the OUT direction flag
        ✓ should support write
        ✓ times out


  30 passing (2s)

```

Oddly, it now takes the node runtime 3-4 seconds to exit after completing the tests compared to master (which instantly exits). I'm not too familiar with the cleanup processing in node, perhaps there are some listeners hanging or other garbage collection going on?

**Comment by kwngo** [comment]
Any ideas why the builds are failing? I tried your branch locally and got the same result.

**Comment by geovie** [comment]
@kwngo It should work for all node version >= v10.16.0 as NAPI_VERSION 4 is required

**Comment by thegecko** [comment]
@geovie did you find any time to update the prebuild steps for this PR?

**Comment by geovie** [comment]
@thegecko Pushed my changes now, it includes the napi prebuild

The `3-4 seconds to exit after completing the tests` are likely due to how `Napi::ThreadSafeFunction` works but it would be good if someone with napi knowledge could have a look over the PR

**Comment by thegecko** [comment]
Thanks @geovie! I'm thinking of bumping the minor version and publishing a pre-release version for testing.

File: .github/workflows/prebuild.yml [hunk 4]
```diff
@@ -32,9 +32,7 @@ jobs:
           submodules: recursive
       - run: npm install
       # Supported Node versions: https://nodejs.org/en/about/releases/
-      - run: npm run prebuild -- -t 8.0.0 -t 10.0.0 -t 12.0.0 -t 14.0.0 -a ${{ matrix.config.arch }}
-      # Supported Electron versions: https://electronjs.org/docs/tutorial/electron-timelines
-      - run: npm run prebuild -- -t 4.0.4 -t 5.0.0 -t 6.0.0 -t 7.0.0 -t 8.0.0 -t 9.0.0 -t 10.0.0 -t 11.0.0 -t 12.0.0-nightly.20201013 -r electron -a ${{ matrix.config.arch }}
+      - run: npm run prebuild -- -a ${{ matrix.config.arch }}
```
Comment (thegecko) [inline .github/workflows/prebuild.yml:35]: To set it up to build N-API, should this read:

```
- run: npm run prebuild -- -t 4 -r napi -a ${{ matrix.config.arch }}
```

The default is `node`

File: package.json [hunk 5]
```diff
@@ -30,8 +30,8 @@
     "url": "https://github.com/tessel/node-usb.git"
   },
   "scripts": {
-    "install": "prebuild-install --verbose || node-gyp rebuild",
-    "prebuild": "prebuild --force --strip --verbose",
+    "install": "prebuild-install --runtime napi || node-gyp rebuild",
```
Comment (thegecko) [inline package.json:33]: Is it worth keeping the `--verbose` flag? Not sure of the benefit it brings to the user.

**Review by thegecko — CHANGES_REQUESTED** [review]
Couple of questions on the prebuild changes

File: .github/workflows/prebuild.yml [hunk 6]
```diff
@@ -32,9 +32,7 @@ jobs:
           submodules: recursive
       - run: npm install
       # Supported Node versions: https://nodejs.org/en/about/releases/
-      - run: npm run prebuild -- -t 8.0.0 -t 10.0.0 -t 12.0.0 -t 14.0.0 -a ${{ matrix.config.arch }}
-      # Supported Electron versions: https://electronjs.org/docs/tutorial/electron-timelines
-      - run: npm run prebuild -- -t 4.0.4 -t 5.0.0 -t 6.0.0 -t 7.0.0 -t 8.0.0 -t 9.0.0 -t 10.0.0 -t 11.0.0 -t 12.0.0-nightly.20201013 -r electron -a ${{ matrix.config.arch }}
```
Comment (thegecko) [inline .github/workflows/prebuild.yml:37]: Does electron support N-API prebuilds?

File: package.json (hunk 5 above)
Comment (geovie) [reply package.json:33]: I'll put it back, didn't remove it on purpose 👍

File: .github/workflows/prebuild.yml (hunk 4 above)
Comment (geovie) [reply .github/workflows/prebuild.yml:35]: `--runtime napi -t 4` is set in the prebuild script in `package.json`

File: .github/workflows/prebuild.yml (hunk 6 above)
Comment (geovie) [reply .github/workflows/prebuild.yml:37]: Yes electron can work with the same prebuild as node

**Comment by thegecko** [comment]
Thanks @geovie 

The N-API version is now published on npm for testing purposes.

To install it:

> npm i usb@testing

**Comment by thegecko** [comment]
@geovie 

The NODE-API build is failing on Apple M1 hardware (darwin/arm64). Any idea how to fix it?

```bash
thegecko@Robs-MacBook-Air node-usb % npm run prebuild                                                 
> usb@1.7.0-alpha.1 prebuild /Users/thegecko/Projects/node-usb
> prebuild --runtime napi -t 4 --force --strip --verbose
prebuild info begin Prebuild version 10.0.1
prebuild info build Preparing to prebuild usb@1.7.0-alpha.1 for napi 4 on darwin-arm64 using node-gyp
prebuild verb starting build process node-gyp 
prebuild verb execute node-gyp with `node index.js rebuild --napi_build_version=4 --target_arch=arm64` 
make: *** No rule to make target `Release/obj.target/libusb/libusb/libusb/core.o', needed by `Release/usb.a'.  Stop.
prebuild ERR! build error 
prebuild ERR! stack Error: `make` failed with exit code: 2
prebuild ERR! stack     at ChildProcess.onExit (/Users/thegecko/Projects/node-usb/node_modules/node-gyp/lib/build.js:194:23)
prebuild ERR! stack     at ChildProcess.emit (events.js:314:20)
prebuild ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:276:12)
prebuild ERR! not ok 
prebuild ERR! build Error: `make` failed with exit code: 2
prebuild ERR! build     at ChildProcess.onExit (/Users/thegecko/Projects/node-usb/node_modules/node-gyp/lib/build.js:194:23)
prebuild ERR! build     at ChildProcess.emit (events.js:314:20)
prebuild ERR! build     at Process.ChildProcess._handle.onexit (internal/child_process.js:276:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! usb@1.7.0-alpha.1 prebuild: `prebuild --runtime napi -t 4 --force --strip --verbose`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the usb@1.7.0-alpha.1 prebuild script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/thegecko/.npm/_logs/2021-04-04T10_28_33_739Z-debug.log
thegecko@Robs-MacBook-Air node-usb % 
```

**Comment by kevinmehall** [comment]
Do you have the libusb submodule checked out? (`git submodule update --init`)

**Comment by thegecko** [comment]
> Do you have the libusb submodule checked out?

Man, this could be embarrassing....

**Comment by thegecko** [comment]
Thanks @kevinmehall, I always forget that step.

Have added a `darwin-arm64` N-API binary to the test release.

**Comment by thegecko** [comment]
**Testing Results**

Electron


Version | 1.x | 2.x | 3.x | 4.x | 5.x | 6.x | 7.x | 8.x | 9.x | 10.x | 11.x | 12.x | 13.x (beta) | 14.x (nightly)
-|-|-|-|-|-|-|-|-|-|-|-|-|-|-
Pass|❌|❌|❌|✅|✅|✅|✅|✅|✅|✅|✅|✅||

**Pull request merged**